packages/store/addon/-private/system/model/internal-model.ts

```diff
@@ -301,6 +320,31 @@ export default class InternalModel {
       }
 
       this._record = store._modelFactoryFor(this.modelName).create(createOptions);
+      if (DEBUG) {
+        let klass = this._record.constructor;
+        let deprecations = lookupDeprecations(klass);
+        [
+          'becameError',
+          'becameInvalid',
+          'didCreate',
+          'didDelete',
+          'didLoad',
+          'didUpdate',
+          'ready',
+          'rolledBack',
+        ].forEach(methodName => {
+          if (typeof this._record[methodName] === 'function') {
```

Comment:
I believe this is where we need to guard this check behind an `instanceof DS.Model` check

Reply:
Although @hjdivad may have other thoughts? The TL;DR here is this deprecation check causes non DS.Model instances (like M3's) to trap this access and treat the methodName as an attribute.

Reply:
with that change i only get 1 failure:
```
not ok 140 Chrome 74.0 - [22 ms] - unit/model: .save errors getting updated via the store and removed upon setting a new value
    ---
        actual: >
            true
        expected: >
            false
        stack: >
                at http://localhost:7357/assets/tests.js:3952:18
                at tryCatcher (http://localhost:7357/assets/vendor.js:77288:21)
                at invokeCallback (http://localhost:7357/assets/vendor.js:77461:33)
                at publish (http://localhost:7357/assets/vendor.js:77447:9)
                at publishRejection (http://localhost:7357/assets/vendor.js:77383:5)
                at http://localhost:7357/assets/vendor.js:71821:16
        message: >
            model done saving
        negative: >
            false
        browser log: |
```

Reply:
> The TL;DR here is this deprecation check causes non DS.Model instances (like M3's) to trap this access and treat the methodName as an attribute.

Checking `instanceof Model` seems ok here.

FWIW M3 models are not `Proxy`s.  The reason the property access is trapped is because Ember is adding the proxy wrapper in dev to catch es5 property access across Ember proxies (via `unknownProperty`).

Reply:
@hjdivad can you explain this more? especially the 'adding the proxy wrapper'?

> > 
> FWIW M3 models are not `Proxy`s. The reason the property access is trapped is because Ember is adding the proxy wrapper in dev to catch es5 property access across Ember proxies (via `unknownProperty`).

Reply:
@pete-the-pete In dev mod Ember will give you back native proxies (and since you're mostly looking in dev you'll mostly see these proxies).

The reason for this is to throw an error if you try to access a property via a native property accessor instead of `Ember.get` when doing so cannot work in production.  I forget the exact rules, but broadly it's if you access a property from an object that has no descriptor, but still has a value computed via `unknownProperty`.

The reason this cannot work in production is twofold.  In the first place, native proxies aren't supported in IE11 and there's no way to polyfill them.  The only reason Ember can support something like a proxy is by requiring an opt-in at every call site (ie `Ember.get`).

The second reason is that there are (potentially?) issues with using `Proxy` in production for this purpose anyway. Here I'm less hazy on the details but a) performance concerns and b) it's not clear that using Proxies in this way is intended usage and it's not clear that doing so won't cause breakage in the future, especially with the introduction of private fields.  This is discussed somewhat in the [tracked properties rfc](https://github.com/emberjs/rfcs/blob/master/text/0410-tracked-properties.md#we-could-wait-on-private-fields-and-proxy-developments) which has links to further discussion.  A word of warning: those links lead to a very deep rabbit hole.

Reply:
🙇 thanks for the good explanation.

I don't mind rabbit holes. 🔦 ⛑